include/deal.II/base/mpi_compute_index_owner_internal.h

```diff
@@ -0,0 +1,799 @@
+                AssertThrowMPI(ierr);
+
+                // unpack the message and translate the dictionary-local
+                // indices coming via MPI to the global index range
+                const types::global_dof_index index_offset =
+                  static_cast<types::global_dof_index>(status.MPI_SOURCE) *
+                  dict.dofs_per_process;
+                unsigned int offset = 0;
+                while (offset < buffer.size())
+                  {
+                    std::vector<std::pair<unsigned, unsigned int>>
+                      &my_intervals = requested_indices[buffer[offset].first];
+                    AssertIndexRange(offset + buffer[offset].second,
+                                     buffer.size());
+
+                    // put the indices into existing ranges from other
+                    // sections of the dictionary. At this stage, we translate
+                    // from the dictionary-local index space to the global one
+                    // and then back into the local index space of the owning
+                    // process
+                    for (unsigned int i = offset + 1;
+                         i < offset + buffer[offset].second + 1;
+                         ++i)
+                      my_intervals.emplace_back(
+                        owned_indices.index_within_set(index_offset +
+                                                       buffer[i].first),
+                        owned_indices.index_within_set(index_offset +
+                                                       buffer[i].second - 1) +
+                          1);
+                    offset += buffer[offset].second + 1;
+                  }
+                AssertDimension(offset, buffer.size());
+              }
+
+            if (send_requests.size() > 0)
+              MPI_Waitall(send_requests.size(),
+                          send_requests.data(),
+                          MPI_STATUSES_IGNORE);
+
+            for (auto &it : requested_indices)
+              {
+                // sort the intervals and merge nearby index ranges. these
+                // appear naturally here because the dictionary might split
+                // some ranges onto two processors
+                std::sort(it.second.begin(), it.second.end());
+                unsigned int last = 0;
+                for (unsigned int check = 1; check < it.second.size(); ++check)
+                  if (it.second[check].first == it.second[last].second)
+                    it.second[last].second =
+                      std::max(it.second[last].second, it.second[check].second);
+                  else
+                    {
+                      ++last;
+                      it.second[last] = it.second[check];
+                    }
+
+                AssertIndexRange(last, it.second.size());
+
+                // erase duplicates
+                if (last + 1 != it.second.size())
+                  it.second.erase(it.second.begin() + last + 1,
```

Comment:
would an IndexSet do all this logic of sorting/merging for you here?

Reply:
I tried, but `IndexSet::add_range` implies quadratic complexity in the number of ranges; for a typical example of an `FE_Q`, we can have thousands of ranges. I tried to work around it by different index sets and extracting/compressing indices, but then the code is not nicer than what I have here. To see the "code overhead" of skipping the index set, have a look at the [fourth commit](https://github.com/dealii/dealii/pull/8772/commits/3f756a1c9cf951405bd1cb96933c707bfaa1346c).

Reply:
Alright, but we might want to rethink the implementation in IndexSet then. Implementing an ``add_ranges()`` function should be able to work in O(nlogn).

Reply:
We already have `add_ranges()` -- it's the function that computes the union of two `IndexSet`s.

Reply:
Right, how could I miss that. Well, maybe because `add_indices` is not the name I was looking for (rather `merge` or something like that). I will go back to the IndexSet solution, given that it makes code slightly simpler.